engine/table/src/main/java/io/deephaven/engine/table/impl/BlinkTableTools.java

```diff
@@ -183,4 +117,134 @@ public static boolean isBlink(Table table) {
+            resultTable.setRefreshing(true);
+            resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+            resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+            resultTable.setFlat();
+
+            if (resultTable.size() < sizeLimit) {
+                resultListener = new BaseTable.ListenerImpl(getDescription(), parent, resultTable) {
+                    @Override
+                    public void onUpdate(TableUpdate upstream) {
+                        BlinkToAppendOnlyOperation.this.onUpdate(upstream);
+                    }
+                };
+            }
+
+            return new Result<>(resultTable, resultListener);
+        }
+
+        private void onUpdate(final TableUpdate upstream) {
+            if (upstream.modified().isNonempty() || upstream.shifted().nonempty()) {
+                throw new IllegalStateException("Blink tables should not modify or shift!");
+            }
+            if (upstream.added().isEmpty()) {
+                return;
+            }
+
+            final TableUpdateImpl downstream = new TableUpdateImpl();
+            downstream.added = appendRows(resultTable.size(), upstream.added(), false);
+            Assert.eqTrue(downstream.added.isNonempty(), "downstream.added.isNonempty()");
+            downstream.modified = RowSetFactory.empty();
+            downstream.removed = RowSetFactory.empty();
+            downstream.modifiedColumnSet = ModifiedColumnSet.EMPTY;
+            downstream.shifted = RowSetShiftData.EMPTY;
+            resultTable.notifyListeners(downstream);
+
+            if (resultTable.size() == sizeLimit) {
+                // No more rows can be appended, so remove the listener and remove all references
+                resultListener.forceReferenceCountToZero();
+                resultListener = null;
+            }
+        }
+
+        private RowSet appendRows(final long currentSize, final RowSet newRows, final boolean usePrev) {
+            long newRowsSize = newRows.size();
+            Assert.gtZero(newRowsSize, "newRowsSize");
+            RowSet rowsToAdd = null;
+            if (currentSize > sizeLimit - newRowsSize) {
+                newRowsSize = (sizeLimit - currentSize);
+                rowsToAdd = newRows.subSetByPositionRange(0, newRowsSize);
+            }
+            final long totalSize = currentSize + newRowsSize;
+            final RowSet newRange = RowSetFactory.fromRange(currentSize, totalSize - 1);
+
+            try (final SafeCloseable ignored = rowsToAdd) {
+                if (rowsToAdd == null) {
+                    rowsToAdd = newRows;
+                }
+                ChunkUtils.copyData(sourceColumns, rowsToAdd, destColumns, newRange, usePrev);
+            }
+            if (resultTable != null) {
+                resultTable.getRowSet().writableCast().insertRange(newRange.firstRowKey(), newRange.lastRowKey());
```

Comment:
I would just do this unconditionally in the listener, instead of in this method. Cleaner, and separates concerns.